### PR TITLE
[serve] Create CurrentState & GoalState

### PR DESCRIPTION
**Pull request by ijrsvt**
## Why are these changes needed?

Incremental progress to making the serve controller goal-state driven. This is a minor refactor and mainly breaks out a new class called `SystemState`. This class keeps all information about Endpoints, Routes & Backends. There is one for the current state (what the cluster currently looks like) and one for the goal state (what the cluster should look like). The ActorStateReconciler class will eventually become more of a pending changes class.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/controller.py [hunk 1]
```diff
@@ -377,29 +395,30 @@ def _recover_actor_handles(self) -> None:
...
         return autoscaling_policies
 
 
 @dataclass
 class Checkpoint:
-    config: ConfigurationStore
+    goal_state: SystemState
+    current_state: SystemState
     reconciler: ActorStateReconciler
+    # TODO(ilr) Rename to PendingState
```
Comment (simon-mo) [inline python/ray/serve/controller.py:421]: rename what to PendingState?

File: python/ray/serve/controller.py [hunk 2]
```diff
@@ -91,9 +91,9 @@ class Config:
 
 
 @dataclass
-class ConfigurationStore:
+class SystemState:
     backends: Dict[BackendTag, BackendInfo] = field(default_factory=dict)
-    traffic_policies: Dict[EndpointTag, TrafficPolicy] = field(
+    traffic_policies: Dict[EndpointTag, Tuple[TrafficPolicy]] = field(
```
Comment (simon-mo) [inline python/ray/serve/controller.py:96]: namedtuple?

**Review by simon-mo — COMMENTED** [review]
Can you add some descriptions to this PR? What's in scope? `goal_state` seems like unused.

File: python/ray/serve/controller.py (hunk 1 above)
Comment (ijrsvt) [reply python/ray/serve/controller.py]: Reconciler, clairifed with comment.

**Review by simon-mo — APPROVED** [review]
LGTM

**Pull request merged**